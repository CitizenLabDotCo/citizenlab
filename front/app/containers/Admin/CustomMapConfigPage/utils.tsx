// getFeatureLayerInitialTitleMultiloc

import { IMapConfig } from 'api/map_config/types';

// Description: This function is used to generate the initial title multiloc for an Esri feature layer.
export const getFeatureLayerInitialTitleMultiloc = (
  layerTitle: string,
  tenantLocales: string[],
  subLayerCount: number
) => {
  const title_multiloc = {};
  const title = // If there are sublayers, append the sublayer count to the title
    subLayerCount > 1 ? `${layerTitle} (${subLayerCount})` : `${layerTitle}`;

  tenantLocales.forEach(
    (tenantLocale) => (title_multiloc[`${tenantLocale}`] = title)
  );
  return title_multiloc;
};

// getLayerType
// Description: Returns the layer type of a MapConfig. All layers in a Map Config will have the same type.
export const getLayerType = (mapConfig: IMapConfig) => {
  const layers = mapConfig.data.attributes.layers;
  return layers[0]?.type;
};
