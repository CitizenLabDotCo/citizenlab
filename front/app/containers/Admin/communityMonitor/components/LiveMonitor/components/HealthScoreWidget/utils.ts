import { colors, Locale } from '@citizenlab/cl2-component-library';

import { CommunityMonitorSentimentScoreAttributes } from 'api/community_monitor_scores/types';

import { QuarterlyScores } from './types';

export const categoryColors = {
  overall: colors.green400,
  quality_of_life: colors.teal400,
  service_delivery: colors.blue400,
  governance_and_trust: colors.brown,
  other: colors.grey700,
};

// transformSentimentScoreData:
// Transform the sentiment score data object into a more usable format with arrays.
export function transformSentimentScoreData(
  scoreResults: CommunityMonitorSentimentScoreAttributes | undefined,
  locale: Locale
): QuarterlyScores | null {
  if (!scoreResults || !scoreResults.overall || !scoreResults.categories) {
    return null;
  }

  // Generate array for overall health scores
  const overallHealthScores = Object.entries(scoreResults.overall.averages).map(
    ([period, score]) => ({ period, score: Number(score) })
  );

  // Generate array for total health score counts
  const totalHealthScoreCounts = Object.entries(
    scoreResults.overall.totals
  ).map(([period, totalsPerSentimentValue]) => ({
    period,
    totals: Object.entries(totalsPerSentimentValue).map(
      ([sentimentValue, count]) => ({ sentimentValue, count: Number(count) })
    ),
  }));

  // Generate array for category health scores
  const categoryHealthScores = Object.entries(
    scoreResults.categories.averages
  ).map(([category, scores]) => {
    const localizedLabel =
      scoreResults.categories?.multilocs[category]?.[locale] || category;

    const scoreEntries = Object.entries(scores).map(([period, score]) => ({
      period,
      score,
    }));

    return {
      category,
      localizedLabel,
      scores: scoreEntries,
    };
  });

  return { overallHealthScores, categoryHealthScores, totalHealthScoreCounts };
}

// getYearFilter:
// Get the current year based on the date. If a year is provided in the URL, use that instead.
export const getYearFilter = (search: URLSearchParams) => {
  const year = search.get('year');
  if (year) {
    return year;
  }

  const date = new Date();
  return date.getFullYear().toString();
};

// getQuarterFilter:
// Get the current quarter based on the date. If a quarter is provided in the URL, use that instead.
export const getQuarterFilter = (search: URLSearchParams) => {
  const quarter = search.get('quarter');
  if (quarter) {
    return quarter;
  }

  const date = new Date();
  return Math.floor(date.getMonth() / 3 + 1).toString();
};
