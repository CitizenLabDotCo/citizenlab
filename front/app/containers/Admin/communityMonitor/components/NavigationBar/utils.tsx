import { FormatMessage, ITab } from 'typings';

import messages from '../../messages';

export const getCommunityMonitorTabs = (
  formatMessage: FormatMessage
): ITab[] => {
  return [
    {
      label: formatMessage(messages.liveMonitor),
      url: `/admin/community-monitor/live-monitor`,
      name: 'live-monitor',
    },
    {
      label: formatMessage(messages.participants),
      name: 'participants',
      url: `/admin/community-monitor/participants`,
    },
    {
      label: formatMessage(messages.reports),
      url: '/admin/community-monitor/reports',
      name: 'reports',
    },
    {
      label: formatMessage(messages.settings),
      name: 'settings',
      url: '/admin/community-monitor/settings',
    },
  ];
};
