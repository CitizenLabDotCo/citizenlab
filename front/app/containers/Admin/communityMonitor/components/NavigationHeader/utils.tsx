import { RouteType } from 'routes';

import clHistory from 'utils/cl-router/history';

import messages from './messages';

export type CommunityMonitorTab =
  | 'live-monitor'
  | 'participants'
  | 'reports'
  | 'settings';

export const TABS: CommunityMonitorTab[] = [
  'live-monitor',
  'participants',
  'reports',
  'settings',
];

// Get the label for a community monitor tab
export const getTabLabel = (tab: CommunityMonitorTab) => {
  switch (tab) {
    case 'live-monitor':
      return messages.liveMonitor;
    case 'participants':
      return messages.participants;
    case 'reports':
      return messages.reports;
    case 'settings':
      return messages.settings;
  }
};

// Navigate to a selected tab
export const navigateToTab = (tab: CommunityMonitorTab, projectId?: string) => {
  const pathMap: Record<CommunityMonitorTab, RouteType> = {
    'live-monitor': '/admin/community-monitor/live-monitor',
    participants: `/admin/community-monitor/participants/projects/${projectId}`,
    reports: '/admin/community-monitor/reports',
    settings: '/admin/community-monitor/settings',
  };
  clHistory.push(pathMap[tab]);
};
