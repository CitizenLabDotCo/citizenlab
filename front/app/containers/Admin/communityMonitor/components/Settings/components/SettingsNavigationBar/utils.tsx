import { FormatMessage, ITab } from 'typings';

import messages from '../../messages';

export const getSettingsTabs = (formatMessage: FormatMessage): ITab[] => {
  return [
    {
      name: 'survey',
      label: formatMessage(messages.surveySettings),
      url: `/admin/community-monitor/settings/survey`,
    },
    {
      name: 'access_rights',
      label: formatMessage(messages.accessRights),
      url: `/admin/community-monitor/settings/access-rights`,
    },
    {
      name: 'moderator_management',
      label: formatMessage(messages.management),
      url: `/admin/community-monitor/settings/moderators`,
    },
  ];
};
