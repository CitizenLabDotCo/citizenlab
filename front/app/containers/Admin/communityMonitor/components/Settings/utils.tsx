export type SettingsTab =
  | 'survey_settings'
  | 'action_form'
  | 'moderator_management';
