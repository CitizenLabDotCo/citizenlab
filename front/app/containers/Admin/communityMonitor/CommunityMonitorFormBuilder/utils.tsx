import React from 'react';

import { Box } from '@citizenlab/cl2-component-library';

import AccessRightsNotice from 'containers/Admin/projects/project/nativeSurvey/AccessRightsNotice';

import { FormBuilderConfig } from 'components/FormBuilder/utils';
import Warning from 'components/UI/Warning';

import { FormattedMessage } from 'utils/cl-intl';

import messages from './messages';

export const communityMonitorConfig: FormBuilderConfig = {
  type: 'survey',
  formBuilderTitle: messages.survey,
  viewFormLinkCopy: messages.viewSurvey,
  toolboxTitle: messages.addSurveyContent,
  formSavedSuccessMessage: messages.successMessage,
  supportArticleLink: messages.supportArticleLink,
  toolboxFieldsToInclude: ['page', 'sentiment_linear_scale'],
  formCustomFields: undefined,
  displayBuiltInFields: false,
  builtInFields: [],
  showStatusBadge: true,
  isLogicEnabled: false,
  alwaysShowCustomFields: true,
  isFormPhaseSpecific: true,
  getWarningNotice: () => {
    return (
      <Box id="e2e-warning-notice" mb="16px">
        <Warning>
          <FormattedMessage {...messages.existingSubmissionsWarning} />
        </Warning>
      </Box>
    );
  },
  getAccessRightsNotice: (projectId, phaseId, handleClose) => {
    return projectId && phaseId ? (
      <AccessRightsNotice
        projectId={projectId}
        phaseId={phaseId}
        handleClose={handleClose}
      />
    ) : null;
  },
};
