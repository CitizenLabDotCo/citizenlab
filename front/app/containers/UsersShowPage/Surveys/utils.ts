import { saveAs } from 'file-saver';

import { API_PATH } from 'containers/App/constants';

import { requestBlob } from 'utils/requestBlob';

export const downloadSurveySubmission = async (id: string) => {
  const apiEndpoint = `${API_PATH}/ideas/${id}/as_xlsx`;
  const fileName = 'survey_response.xlsx';

  const blob = await requestBlob(
    apiEndpoint,
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
  );
  saveAs(blob, fileName);
};
