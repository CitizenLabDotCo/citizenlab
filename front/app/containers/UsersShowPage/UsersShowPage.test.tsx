// @ts-nocheck
// libraries
import React from 'react';
import { shallow } from 'enzyme';
import { intl } from 'utils/cl-intl';
import useUser from 'hooks/useUser';

// component to test
import { UsersShowPage } from './';

// mock utilities
jest.mock('utils/cl-intl');
jest.mock('services/comments');
jest.mock('services/ideas');
jest.mock('services/users');
jest.mock('modules', () => ({ streamsToReset: [] }));
jest.mock('components/Outlet', () => 'Outlet');

const mockScroll = jest.fn();
global.scrollTo = mockScroll;

const user = makeUser().data;
jest.mock('hooks/useUser', () => jest.fn());

import { makeUser } from 'services/__mocks__/users';

describe('<UsersShowPage />', () => {
  it('renders correctly by default', () => {
    useUser.mockImplementation(() => user);
    const wrapper = shallow(
      <UsersShowPage params={{ slug: user.attributes.slug }} intl={intl} />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly on the other tab', () => {
    useUser.mockImplementation(() => user);
    const wrapper = shallow(
      <UsersShowPage params={{ slug: user.attributes.slug }} intl={intl} />
    );
    wrapper.find('WrappedUserNavbar').prop('selectTab')('comments')();
    wrapper.update();
    expect(mockScroll).toHaveBeenCalledWith(0, 0);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly with an erroneous user', () => {
    useUser.mockImplementation(() => new Error());
    const wrapper = shallow(
      <UsersShowPage params={{ slug: 'non-existent' }} intl={intl} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
