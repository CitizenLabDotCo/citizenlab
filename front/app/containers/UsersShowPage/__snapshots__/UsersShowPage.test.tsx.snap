// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersShowPage /> renders correctly by default 1`] = `
<Fragment>
  <WrappedUsersShowPageMeta
    user={
      {
        "attributes": {
          "avatar": {
            "large": null,
            "medium": null,
            "small": null,
          },
          "bio_multiloc": {},
          "confirmation_required": false,
          "created_at": "2018-11-26T15:41:19.782Z",
          "custom_field_values": {
            "birthyear": 1990,
            "domicile": "outside",
            "gender": "female",
            "politician": "retired_politician",
          },
          "email": "test@citizenlab.co",
          "first_name": "Test",
          "highest_role": "user",
          "invite_status": null,
          "last_name": "Citizenlab",
          "locale": "en",
          "registration_completed_at": "2018-11-26T15:40:54.355Z",
          "roles": [],
          "slug": "test-citizenlab",
          "unread_notifications": 0,
          "updated_at": "2018-11-26T15:41:19.782Z",
        },
        "id": "522ae8cc-a5ed-4d31-9aa0-470904934ec6",
        "relationships": {
          "granted_permissions": {
            "data": [],
          },
        },
        "type": "user",
      }
    }
  />
  <UsersShowPage__Container
    id="e2e-usersshowpage"
  >
    <UserHeader
      userSlug="test-citizenlab"
    />
    <WrappedUserNavbar
      currentTab="ideas"
      selectTab={[Function]}
      userId="522ae8cc-a5ed-4d31-9aa0-470904934ec6"
    />
    <UsersShowPage__StyledContentContainer
      maxWidth={1166}
    >
      <UsersShowPage__UserIdeas>
        <Memo()
          authorId="522ae8cc-a5ed-4d31-9aa0-470904934ec6"
          invisibleTitleMessage={
            {
              "defaultMessage": "All posts submitted by this participant",
              "id": "app.containers.UsersShowPage.invisibleTitlePostsList",
            }
          }
          type="load-more"
        />
      </UsersShowPage__UserIdeas>
    </UsersShowPage__StyledContentContainer>
  </UsersShowPage__Container>
</Fragment>
`;

exports[`<UsersShowPage /> renders correctly on the other tab 1`] = `
<Fragment>
  <WrappedUsersShowPageMeta
    user={
      {
        "attributes": {
          "avatar": {
            "large": null,
            "medium": null,
            "small": null,
          },
          "bio_multiloc": {},
          "confirmation_required": false,
          "created_at": "2018-11-26T15:41:19.782Z",
          "custom_field_values": {
            "birthyear": 1990,
            "domicile": "outside",
            "gender": "female",
            "politician": "retired_politician",
          },
          "email": "test@citizenlab.co",
          "first_name": "Test",
          "highest_role": "user",
          "invite_status": null,
          "last_name": "Citizenlab",
          "locale": "en",
          "registration_completed_at": "2018-11-26T15:40:54.355Z",
          "roles": [],
          "slug": "test-citizenlab",
          "unread_notifications": 0,
          "updated_at": "2018-11-26T15:41:19.782Z",
        },
        "id": "522ae8cc-a5ed-4d31-9aa0-470904934ec6",
        "relationships": {
          "granted_permissions": {
            "data": [],
          },
        },
        "type": "user",
      }
    }
  />
  <UsersShowPage__Container
    id="e2e-usersshowpage"
  >
    <UserHeader
      userSlug="test-citizenlab"
    />
    <WrappedUserNavbar
      currentTab="ideas"
      selectTab={[Function]}
      userId="522ae8cc-a5ed-4d31-9aa0-470904934ec6"
    />
    <UsersShowPage__StyledContentContainer
      maxWidth={1166}
    >
      <UsersShowPage__UserIdeas>
        <Memo()
          authorId="522ae8cc-a5ed-4d31-9aa0-470904934ec6"
          invisibleTitleMessage={
            {
              "defaultMessage": "All posts submitted by this participant",
              "id": "app.containers.UsersShowPage.invisibleTitlePostsList",
            }
          }
          type="load-more"
        />
      </UsersShowPage__UserIdeas>
    </UsersShowPage__StyledContentContainer>
  </UsersShowPage__Container>
</Fragment>
`;

exports[`<UsersShowPage /> renders correctly with an erroneous user 1`] = `
<UsersShowPage__NotFoundContainer
  className=""
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <meta
      content="404"
      name="prerender-status-code"
    />
  </HelmetWrapper>
  <p>
    We couldn't find this user's profile. If it was here before, it may have been deleted.
  </p>
  <ButtonWrapper
    icon="arrow-left"
    linkTo="/"
    text="Go back to the previous page"
  />
</UsersShowPage__NotFoundContainer>
`;
